File: src/engine/rehydrator.py
"""
Batch rehydrator: replaces placeholders back to original values.

Used for non-streamed (complete) response bodies.
For streaming, see streaming.py.
"""

from __future__ import annotations

import re

from .entities import RehydrateResult

# Matches any typed positional placeholder: [ENTITY_TYPE_N]
_PLACEHOLDER_RE = re.compile(r"\[([A-Z_]+)_(\d+)\]")


class BatchRehydrator:
    def rehydrate(
        self,
        text: str,
        placeholder_map: dict[str, str],
    ) -> tuple[str, int]:
        """
        Replace all placeholders in text with their original values.

        Returns:
            (rehydrated_text, number_of_substitutions)
        """
        if not placeholder_map or not text:
            return text, 0

        count = 0

        def _replace(m: re.Match[str]) -> str:
            nonlocal count
            placeholder = m.group(0)
            original = placeholder_map.get(placeholder)
            if original is not None:
                count += 1
                return original
            return placeholder  # Unknown placeholder — pass through unchanged.

        result = _PLACEHOLDER_RE.sub(_replace, text)
        return result, count

File: crates/engine-core/src/rehydrator.rs
/*!
Batch rehydrator: replaces placeholders back to original values in a complete string.

For streaming (chunk-by-chunk) use `StreamingRehydrator` in `streaming.rs`.
*/

use std::collections::HashMap;

use once_cell::sync::Lazy;
use regex::Regex;

// Matches any typed positional placeholder: [ENTITY_TYPE_N]
static PLACEHOLDER_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"\[[A-Z_]+_\d+\]").unwrap());

pub struct BatchRehydrator;

impl BatchRehydrator {
    /// Replace all known placeholders in `text`.
    ///
    /// Returns the rehydrated text and the count of substitutions made.
    pub fn rehydrate(&self, text: &str, map: &HashMap<String, String>) -> (String, usize) {
        if map.is_empty() || text.is_empty() {
            return (text.to_string(), 0);
        }

        let mut count = 0usize;
        let result = PLACEHOLDER_RE.replace_all(text, |caps: &regex::Captures<'_>| {
            let ph = caps.get(0).unwrap().as_str();
            if let Some(original) = map.get(ph) {
                count += 1;
                original.clone()
            } else {
                ph.to_string()
            }
        });

        (result.into_owned(), count)
    }
}
